webkitbot fails to run due to Autoinstall issues

lingcherd opened the pull request:
#### c0c7953c13f3ba6d0b9e4a365ffd55c6cb4e125f
<pre>
webkitbot fails to run due to Autoinstall issues
<a href="https://bugs.webkit.org/show_bug.cgi?id=262560">https://bugs.webkit.org/show_bug.cgi?id=262560</a>
rdar://116135662 (webkitbot intermittently failing due to networking issues(?))

Reviewed by Alexey Proskuryakov.

Populate PROXY environment variables to webkit-patch to allow download of
PIP packages by Autoinstall, and launch webkit-patch by explicitly calling
python3 so sys.executable get populated in Autoinstall.
Updated HttpsProxyAgent statement to fix &quot;httpsProxyAgent is not a constructor&quot; error.

* Tools/WebKitBot/src/WebKitBot.mjs:

Canonical link: <a href="https://commits.webkit.org/268855@main">https://commits.webkit.org/268855@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/baf1f18e7954d82b6c48677f6b1ab068b8979d86

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20836 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21243 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21903 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22725 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19416 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21072 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24482 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21419 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20712 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21057 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20830 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18093 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23579 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17994 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18906 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25205 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19074 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19088 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23114 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19667 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16705 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18909 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5008 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23238 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19483 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Tools/WebKitBot/src/WebKitBot.mjs:
```diff
@@ -471,7 +471,12 @@ Type \`help COMMAND\` for help on my individual commands.`,
             const webkitPatchPath = path.resolve("BotWebKit", "Tools", "Scripts", "webkit-patch");
-            results = await execFileAsync(webkitPatchPath, [
+            var pythonPath = which.sync('python3', { nothrow: true })
+            if (!pythonPath) {
```
I don't think that we need fallback here - it worked fine without a fallback before (via `/usr/bin/env` inside the script).

lingcherd replied:
If I don't have a fallback, would execFileAsync still work with an empty pythonPath? And even if it does, the Python sys.executable would end up with an empty string and break Autoinstall, wouldn't it?

aproskuryakov replied:
Probably best to also remove `nothrow: true`, and have it fail. But it won't fail in this configuration, because we know that Python is in the path.

Having a manual fallback is dangerous, because it could introduce mysterious behavior if something changes. For a service like this, it's better to fail quickly and loudly in this case.

gsnedders commented on Tools/WebKitBot/src/WebKitBot.mjs:
```diff
@@ -471,7 +472,9 @@ Type \`help COMMAND\` for help on my individual commands.`,
             const webkitPatchPath = path.resolve("BotWebKit", "Tools", "Scripts", "webkit-patch");
-            results = await execFileAsync(webkitPatchPath, [
+            var pythonPath = which.sync('python3')
+            results = await execFileAsync(pythonPath, [
```
What's the reason to invoke like this, rather than invoking it via `#!/usr/bin/env python3`? Presumably we're finding the same `python3` interpreter either way, as we'll have the same `PATH` either way, no?

gsnedders replied:
Oh, `sys.executable`. That should absolutely, definitely, be populated the exact same either way?

lingcherd replied:
sys.executable is not getting populated when BotWebKit is invoked directly using execFileAsync.

aproskuryakov replied:
Right. Also, setting `argv0` in options doesn't work, even though it definitely seems like it should.

Python appears to be doing some illogical things here.

gsnedders replied:
I wonder why, I wonder what `execFileAsync` is doing such that CPython is unable to determine the path of the interpreter…

aproskuryakov replied:
I doubt that it could do much interesting on Node.js side. This general kind of problem appears to be common with Python, see https://bugs.webkit.org/show_bug.cgi?id=225172 for an earlier instance unrelated to Node.

lingcherd commented on Tools/WebKitBot/src/WebKitBot.mjs:
```diff
@@ -26,8 +26,9 @@
 import util from "util";
+import which from "which";
 import {execFile, spawn} from "child_process";
-import HttpsProxyAgent from "https-proxy-agent";
```
This is failing in newer version of Node.

lingcherd committed:
webkitbot fails to run due to Autoinstall issues
https://bugs.webkit.org/show_bug.cgi?id=262560
rdar://116135662 (webkitbot intermittently failing due to networking issues(?))

Reviewed by Alexey Proskuryakov.

Populate PROXY environment variables to webkit-patch to allow download of
PIP packages by Autoinstall, and launch webkit-patch by explicitly calling
python3 so sys.executable get populated in Autoinstall.
Updated HttpsProxyAgent statement to fix "httpsProxyAgent is not a constructor" error.

* Tools/WebKitBot/src/WebKitBot.mjs:

Canonical link: https://commits.webkit.org/268855@main
